synapse/event_auth.py

```diff
@@ -277,11 +277,15 @@ def _is_membership_change_allowed(
 
     key = (EventTypes.JoinRules, "")
     join_rule_event = auth_events.get(key)
+    logger.error("join_rule_event ggre %s", join_rule_event)
     if join_rule_event:
         join_rule = join_rule_event.content.get("join_rule", JoinRules.INVITE)
     else:
         join_rule = JoinRules.INVITE
 
+    # TODO: remove. The auth_rules don't have the join_rules for the given even we're trying auth
+    join_rule = JoinRules.PUBLIC
```

Comment:
In the `/backfill` case, somehow, the `join_rule` `auth_event` is passed in as expected:

```
2021-07-21 08:03:58,583 - synapse.handlers.federation - 2769 - INFO - GET-19 - auth_events contains unknown events: {'$dhCaJSx7qeIOaxj4JYuR3gye48l0m5zuJtAignvju7c'}
...
2021-07-21 08:03:58,595 - synapse.handlers.federation - 2651 - ERROR - GET-19 - _check_event_auth before event_id=$dhCaJSx7qeIOaxj4JYuR3gye48l0m5zuJtAignvju7c auth_events={('m.room.create', ''): <FrozenEventV3 event_id='$6y6T_Bv5nM61HM95XHFa69aBsM_0bIHREhCdyhjUBrg', type='m.room.create', state_key=''>, ('m.room.power_levels', ''): <FrozenEventV3 event_id='$UAItBT9ewW7NOlfrBnRp-T6d-u55K_pcVrsFMDjTEjs', type='m.room.power_levels', state_key=''>, ('m.room.join_rules', ''): <FrozenEventV3 event_id='$926NzxK3igNhvAkSat6en-c1zcw6bQi1H6TbtqaZepA', type='m.room.join_rules', state_key=''>}

```

But when trying to reproduce this with `on_receive_pdu`, it's not coming through:
```
2021-07-21 03:18:32-0500 [-] 2021-07-21 03:18:32,158 - synapse.handlers.federation - 2687 - INFO - receive_pdu - auth_events contains unknown events: {'$AaMWWGt6khyIH93H7HXHoAbl9YWwBv2WssYCz1z-gbM'}
...
2021-07-21 03:18:32-0500 [-] 2021-07-21 03:18:32,160 - synapse.handlers.federation - 2569 - ERROR - receive_pdu - _check_event_auth before event_id=$AaMWWGt6khyIH93H7HXHoAbl9YWwBv2WssYCz1z-gbM auth_events={('m.room.create', ''): <FrozenEventV3 event_id='$PnSiSdv6o2PBPHCPxz4NUJP8u7m-GseAjKt6uiK4vF4', type='m.room.create', state_key=''>, ('m.room.power_levels', ''): <FrozenEventV3 event_id='$DnqfASe-gJXO-5X6rtDvKRwC4-VwJsMOsVcm9a23QAY', type='m.room.power_levels', state_key=''>}
```

Reply:
It looks like this stems from a problem where [`GET /_matrix/federation/v1/event_auth/{roomId}/{eventId}`](https://matrix.org/docs/spec/server_server/r0.1.4#get-matrix-federation-v1-event-auth-roomid-eventid) returns different types of events from `_event_auth_handler.compute_auth_events(...)`.

[`GET /_matrix/federation/v1/event_auth/{roomId}/{eventId}`](https://matrix.org/docs/spec/server_server/r0.1.4#get-matrix-federation-v1-event-auth-roomid-eventid) returns `m.room.create`, `m.room.power_levels`, `m.room.join_rules`, and `m.room.member` events

Whereas `compute_auth_events` will only return `m.room.create`, `m.room.power_levels`, and `m.room.member` (missing the join rules)